documentation/specs/proposed/WASI-tasks.md

```diff
@@ -0,0 +1,182 @@
+- In principle it's possible to compile to .wasm with a few Exec tasks too.
+
+We can make this more user friendly.
+
+### Utility for MSBuild
+- WASI tasks are sandboxed
+- Easier interoperability outside of C#
+- WASI executables are packaged with no outside dependencies 
+- Users from other languages
+
+## Goals for the WASI task feature
+1. specify interface for writing tasks in other languages
+2. running Wasm 
+3. Rust demo task
+
+
+### Steps (???)
+1. figure out the best options for hosting a WASI sandbox from .NET (probably p/invoke into some Rust thing)
+2. write a TaskHost that uses it
+3. write a WASM interface spec to map to the task model
+4. do the invocation stuff
+5. demos
+
+### Prototype features
+- [ ] WasmExec class extending ToolTask taking a .wasm file as a parameter - just runs the file with wasmtime
+- [ ] WasmTask class extending ToolTask taking a .wasm file as a parameter
+    - [ ] Specification for what should this .wasm file export and how it will be ran
+    - [ ] Rust example
+    - [ ] Possibly needs a TaskHost TBD
+
+### Advanced features
+- [ ] RunWasmCode task which takes a .wasm file which exports a function and parameterized to which function to run
+- [ ] Wasm code inside XML 
+    - [ ] WasmCodeFactory 
+- [ ] C# code using wasm exported functions inside xml
+- [ ] Wasm-supported language code inside XML (possibly too ambitious/flaky) [open question](#open-questions-and-considerations)
+    - would need to include the tooling for getting wasm out of the languages which is under heavy development 
+
+
+
+## Design
+### diagram
+
+![diagram](wasi-diagram.svg)
+
+### Choosing WASI runtime [(open questions)](#open-questions-and-considerations)
+https://wasi.dev/ mentions several possible runtimes: Wasmtime, WAMR, WasmEdge, wazero, Wasmer, wasmi, and wasm3.
+An important criterion is popularity/activity in development as the WASM standard is evolving and needs a lot of developers to implement it.
+This leads to considering [Wasmtime](https://wasmtime.dev/) or [Wasmer](https://wasmer.io/).
+Interaction with C# is especially important for us so we will use **Wasmtime** because the integration via a NuGet package is more up to date and there is more active development in tooling and other dotnet projects use it. 
+
+### Testing
+- **TBD**
+- E2E tests - building projects in different languages most important
+
+<!-- Integration tests for logging -->
+<!-- mirror MSBuild\src\Build.UnitTests\BackEnd\TaskHost_Tests.cs --> 
+
+## User Experience
+API should be clear and the Rust task provide an example of how to implement it.
+Then the user adds the task to their .proj file and it runs and logs as if it were a C# task.
+
+## Development remarks (in-progress)
+
+### TODO for this doc
+- create in depth explanations for Wasm/WASI and how its concepts map to MSBuild concepts
+- discuss with people who understand MSBuild internals, WASI and dotnet interaction, users of MSBuild
+
+### Open questions and considerations 
```

Comment:
Today, MSBuild targets can have `Inputs=` and `Outputs=` lists that define what files are read and written by the target, so that we can skip the target if all of the outputs are up-to-date with all of the inputs. Today that's not passed to targets at all. What I'm thinking though is: what if sandboxed tasks inside a target were _only allowed to access those files_? Then incremental behavior would be guaranteed correct, and it'd also be much easier to audit what files were accessed.

This comment was thinking to that time: you'll need to understand the sandboxing on the WASI side and figure out what info you need to provide on the MSBuild engine side to activate it.

Reply:
That's the default behavior of `wasmtime` (CLI): restrict all resources; and all visible files/dirs need to be configured. 
specifying Inputs and Outputs is an important requirement 👍
I expect that to be easy in the API.

Reply:
Great! The MSBuild side will require some new design work since it's not information we carry today.